devito/mpi/distributed.py

```diff
@@ -339,14 +288,98 @@ def all_ranges(self):
         """The global ranges of all MPI ranks."""
         ret = []
         for i in self.all_numb:
-            ret.append(EnrichedTuple(*[range(min(j), max(j) + 1) for j in i],
+            # In cases where ranks are empty, return a range(0, 0)
+            ret.append(EnrichedTuple(*[range(min(j), max(j) + 1) if len(j) > 0
+                                       else range(0, 0) for j in i],
                                      getters=self.dimensions))
         return tuple(ret)
 
+    @cached_property
+    def _obj_comm(self):
+        """An Object representing the MPI communicator."""
+        return MPICommObject(self.comm)
+
+    @cached_property
+    def _obj_neighborhood(self):
+        """
+        A CompositeObject describing the calling MPI rank's neighborhood
+        in the decomposed grid.
+        """
+        return MPINeighborhood(self.neighborhood)
+
+
+class Distributor(DenseDistributor):
+
+    """
+    Decompose a grid over a set of MPI processes.
+
+    Parameters
+    ----------
+    shape : tuple of ints
+        The shape of the domain to be decomposed.
+    dimensions : tuple of Dimensions
+        The domain Dimensions.
+    comm : MPI communicator, optional
+        The set of processes over which the domain is distributed. Defaults to
+        MPI.COMM_WORLD.
+    """
+
+    def __init__(self, shape, dimensions, input_comm=None, topology=None):
+        super().__init__(shape, dimensions)
+
+        if configuration['mpi']:
+            # First time we enter here, we make sure MPI is initialized
+            devito_mpi_init()
+
+            # Note: the cloned communicator doesn't need to be explicitly freed;
+            # mpi4py takes care of that when the object gets out of scope
```

Comment:
Is this really true? Or do you mean that is this only true _specifically_ for cloned comms?
I'd say reference the docs, but...

Reply:
Presumably so? Note that these changes are preexisting code being moved around within the file, rather than new code